core/src/main/java/org/elasticsearch/search/collapse/CollapseBuilder.java

```diff
@@ -74,14 +98,30 @@ public CollapseBuilder(String field) {
     public CollapseBuilder(StreamInput in) throws IOException {
         this.field = in.readString();
         this.maxConcurrentGroupRequests = in.readVInt();
-        this.innerHit = in.readOptionalWriteable(InnerHitBuilder::new);
+        if (in.readBoolean()) {
+            int numInnerHits = in.readVInt();
+            this.innerHits = new ArrayList<>(numInnerHits);
+            for (int i = 0; i < numInnerHits; i++) {
+                innerHits.add(new InnerHitBuilder(in));
+            }
+        }
```

Comment:
Can you use StreamInput#readList ?
You need to check for the version here since this code can receive requests from nodes in previous version.
Something like ````if (in.getVersion.onOrAfter(Version.V_6....)````